lib/si-layer-cache/src/hybrid_cache.rs

```diff
@@ -0,0 +1,174 @@
+use foyer::{
+    DirectFsDeviceOptions, Engine, FifoPicker, HybridCache, HybridCacheBuilder, LargeEngineOptions,
+    RateLimitPicker, RecoverMode,
+};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use telemetry::tracing::error;
+
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+
+use crate::db::serialize;
+use crate::error::LayerDbResult;
+use crate::LayerDbError;
+
+const DEFAULT_DISK_CACHE_RATE_LIMIT: usize = 1024 * 1024 * 1024;
+const DEFAULT_DISK_BUFFER_SIZE: usize = 1024 * 1024 * 128; // 128mb
+const DEFAULT_DISK_BUFFER_FLUSHERS: usize = 2;
+const DEFAULT_DISK_CAPACITY: usize = 1024 * 1024 * 1024 * 16; // 16gb
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+enum MaybeDeserialized<V>
+where
+    V: Serialize + Clone + Send + Sync + 'static,
+{
+    RawBytes(Vec<u8>),
+    DeserializedValue(V),
+}
+
+#[derive(Clone, Debug)]
+pub struct Cache<V>
+where
+    V: Serialize + DeserializeOwned + Clone + Send + Sync + 'static,
+{
+    cache: HybridCache<Arc<str>, MaybeDeserialized<V>>,
+}
+
+impl<V> Cache<V>
+where
+    V: Serialize + DeserializeOwned + Clone + Send + Sync + 'static,
+{
+    pub async fn new(config: CacheConfig) -> LayerDbResult<Self> {
+        let cache: HybridCache<Arc<str>, MaybeDeserialized<V>> = HybridCacheBuilder::new()
+            .memory((config.memory as f32 * config.memory_percentage) as usize)
+            .with_weighter(|_key: &Arc<str>, value: &MaybeDeserialized<V>| size_of_val(value))
+            .storage(Engine::Large)
+            .with_admission_picker(Arc::new(RateLimitPicker::new(
+                config.disk_admission_rate_limit,
+            )))
+            .with_device_options(
+                DirectFsDeviceOptions::new(config.disk_path).with_capacity(config.disk_capacity),
+            )
+            .with_large_object_disk_cache_options(
+                LargeEngineOptions::new()
+                    .with_flushers(config.disk_buffer_flushers)
+                    .with_buffer_pool_size(config.disk_buffer_size)
+                    .with_eviction_pickers(vec![Box::<FifoPicker>::default()]),
+            )
+            .with_recover_mode(RecoverMode::Quiet)
+            .build()
+            .await
+            .map_err(|e| LayerDbError::Foyer(e.into()))?;
+
+        Ok(Self { cache })
+    }
+
+    pub async fn get(&self, key: &str) -> Option<V> {
+        match self.cache.obtain(key.into()).await {
+            Ok(Some(entry)) => match entry.value() {
+                // todo: bad clone here
```

Comment:
I believe all our inner types here are Arcs? So the clone is cheap

Reply:
Yeah, I think that's the case in our specific impl, that way a 30mb snapshot stays as a single 30mb hunk of heap memory and not `n` copies over time.